### ISSUE TITLE
Objective functions do not correctly fill SPAM penalty jacobian

### ISSUE THREAD

**Issue by robpkelly** [pyGSTio/pyGSTi#80]
(sorry in advance for abusing math terminology)

**Describe the bug**
In `pygsti.baseobjs.objectivefn` see [`_spam_penalty_jac_fill`](https://github.com/pyGSTio/pyGSTi/blob/develop/packages/pygsti/baseobjs/objectivefns.py#L1045). I'm pretty sure that the assignments [here](https://github.com/pyGSTio/pyGSTi/blob/develop/packages/pygsti/baseobjs/objectivefns.py#L1125) should be adding another term to the first axis index (possibly the key for `effectvec`?) because as it is each iteration of the loop will overwrite the same row. Additional space is allocated in the ObjectiveFunction `__init__` but never written to, so the returned jacobian will have garbage data at the end!

**Additional context**
See [this failed build](https://travis-ci.org/pyGSTio/pyGSTi/jobs/559008490#L3040) from the test refactor branch, and compare to [this subsequent build](https://travis-ci.org/pyGSTio/pyGSTi/jobs/559038879#L610). In `custom_leastsq`, one uninitialized row of the jacobian meant there was a garbage value in the diagonal, which would rarely (and nondeterministically) cause `mu` to grow very very fast and fill the diagonal with `inf`s, causing the error. The only change between builds was adding a CI step to print out the build environment so I could debug the failed test :rofl:

**Comment by robpkelly** [comment]
Here's a script to demo the bug. Sorry, this one was hard to find a real minimal example for:

```python
import numpy as np

from pygsti.construction import std1Q_XYI as std
from pygsti.baseobjs import objectivefns, profiler

# build objective function
model = std.target_model()
circuits = std.germs

evTree, wrtBlkSize, _, lookup, outcomes_lookup = model.bulk_evaltree_from_resources(
    circuits, distributeMethod='deriv', subcalls=['bulk_fill_probs', 'bulk_fill_dprobs']
)

opLabelAliases = None
cptp_penalty_factor = 0
spam_penalty_factor = 1.0
cntVecMx = totalCntVec = np.ones(evTree.num_final_elements(), 'd')
radius = 1e-4
minProbClip = 1e-4
probClipInterval = (-1e2, 1e2)
gthrMem = None
forcefn_grad = None
poissonPicture = True

objective = objectivefns.LogLFunction(
    model, evTree, lookup, circuits, opLabelAliases,
    cptp_penalty_factor, spam_penalty_factor, cntVecMx, totalCntVec,
    minProbClip, radius, probClipInterval, wrtBlkSize, gthrMem,
    forcefn_grad, poissonPicture, profiler=profiler.DummyProfiler()
)

# space for the objective jacobian is preallocated but uninitialized
# and filled in by the operation; we'll initialize it with something
# conspicuous to make the issue obvious.
objective.jac = np.ones_like(objective.jac) * 8675309

# compute jacobian of arbitrary argument
J = objective.jfn(np.random.randn(60))

# we expect the preallocated space to have been completely overwritten...
assert (J != 8675309).all()
# ... but we're wrong: the last row was never computed!
```

**Comment by enielse** [comment]
Found and fixed this bug in commit 3a2dfa670096ec99883d02449df193065222a675.  The next major release (probably v0.9.9) will have this fix in it.  Since not many people, if anyone, uses the spam penalty factors it doesn't seem worth it to hot-fix master with this.  Let's leave this issue open until the fix *is* incorporated into a release though.

**Comment by enielse** [comment]
This has been fixed in version 0.9.9.
